### PR TITLE
Improve `_fft_convolve`

### PR DESCRIPTION
**Pull request by asi1024**
Rel. #3556.
Blocked by #3571.

Previous:
```
convolve            :    CPU: 1598.203 us   +/-196.056 (min: 1553.705 / max: 5673.330) us     GPU-0: 1607.828 us   +/-198.819 (min: 1562.784 / max: 5725.760) us
convolve            :    CPU: 7661.681 us   +/-683.048 (min: 7492.259 / max:14718.648) us     GPU-0: 7673.338 us   +/-684.819 (min: 7503.136 / max:14733.984) us
convolve            :    CPU: 7556.341 us   +/-45.591 (min: 7495.102 / max: 7804.286) us     GPU-0: 7567.647 us   +/-45.571 (min: 7506.176 / max: 7815.456) us
convolve            :    CPU: 9845.364 us   +/-53.787 (min: 9777.471 / max:10081.744) us     GPU-0: 9856.696 us   +/-53.819 (min: 9788.608 / max:10093.184) us
```
This PR (https://github.com/cupy/cupy/pull/3560/commits/b570a128a6bf1a76136fedc5ba2b7abcc763f43e):
```
convolve            :    CPU: 1428.156 us   +/-157.454 (min: 1391.433 / max: 5490.528) us     GPU-0: 1435.945 us   +/-158.645 (min: 1399.072 / max: 5529.664) us
convolve            :    CPU: 1511.592 us   +/-144.757 (min: 1472.319 / max: 4345.249) us     GPU-0: 1519.854 us   +/-147.089 (min: 1480.000 / max: 4380.480) us
convolve            :    CPU: 1505.722 us   +/-32.617 (min: 1474.274 / max: 1744.067) us     GPU-0: 1513.800 us   +/-32.601 (min: 1482.432 / max: 1752.128) us
convolve            :    CPU: 1509.725 us   +/-72.535 (min: 1474.584 / max: 3164.543) us     GPU-0: 1518.026 us   +/-75.407 (min: 1482.656 / max: 3254.400) us
```

benchmark script:
```py
import numpy

import cupy
import cupyx

for a_len in [10, 200000]:
    for v_len in [10, 200000]:
        a = cupy.random.rand(a_len)
        v = cupy.random.rand(v_len)
        perf = cupyx.time.repeat(
            cupy.convolve, (a, v, 'same'),
            n_warmup=5, max_duration=1)
        print(perf)
```

### CODE REVIEW

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
The PR pads the size to be the closest power of 2 so the fft algorithms can work more efficiently.

**Comment by asi1024** [comment]
Jenkins, test this please.

File: cupy/math/misc.py [hunk 1]
```diff
@@ -65,13 +65,13 @@ def _fft_convolve(a1, a2, mode):
 
     dtype = cupy.result_type(a1, a2)
```
Comment (leofang) [inline cupy/math/misc.py:66]: What if we upcast `a1` and `a2` to the same `dtype` before calling fft?

Comment (leofang) [reply cupy/math/misc.py:66]: I guess it would not matter, as only the non-contiguous test failed...

**Comment by asi1024** [comment]
`cupy-py35` which uses cuFFT 10.1.2 has passed, and `cupy-py3` which uses cuFFT 10.0.0 has failed.

**Comment by grlee77** [comment]
@emcastillo 
Good idea to update the size to a favorable one.

Note, that you don't necessarily need to increase to a power of two, but could instead go to the next smallest number that has only 2, 3, 5, and 7 as its prime factors (ref: https://docs.nvidia.com/cuda/cufft/index.html#accuracy-and-performance). 

SciPy has a utility function for this ([scipy.fft.next_fast_len](https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.next_fast_len.html)), which has a docstring example where the next power of two makes the signal unnecessarily large. pocketfft's prime factors are slightly different from cuFFT's factors: they are {2, 3, 5, 7, 11} for complex-to-complex transforms, but {2, 3, 5} for transforms involving real input or output. The latter would work fine for cuFFT (just neglecting some potential cases involving factors of 7)

Specifically, if scipy >= 1.4 is available, one can call

```Python
fast_size = scipy.fft.next_fast_len(size, real=True)
```
For older SciPy, FFTPACK only supported factors (2, 3, 5) so
```Python
fast_size = scipy.fftpack.next_fast_len(size)
```
would work instead.

If you really wanted to make a cuFFT-dedicated C++ function also supporting factor 7 you could start from the C++ ones in pocketfft as here:
https://github.com/scipy/scipy/blob/3f11e32f3d07e6035b46f77eafec4359875a0416/scipy/fft/_pocketfft/pocketfft_hdronly.h#L411-L436

**Comment by leofang** [comment]
> If you really wanted to make a cuFFT-dedicated C++ function also supporting factor 7 you could start from the C++ ones in pocketfft 

I once thought about adding `cupyx.scipy.fft*.next_fast_len()`, but I didn't have strong motivation (until now seeing this). But, due to the difference between pocketfft and cuFFT, the returned values won't be identical. Not sure if this is acceptable.

**Comment by grlee77** [comment]
> But, due to the difference between pocketfft and cuFFT, the returned values won't be identical. Not sure if this is acceptable.

I would personally find it acceptable, as it is conceptually the same: "find me the next smallest length that will not have poor performance in the underlying FFT library". Particularly in the context of a round-trip transform as for FFT-based convolution, I don't think the user will care what intermediate size was used internally.

**Comment by leofang** [comment]
Hi @grlee77 You know I almost always agree with you on performance considerations 😁 I was refreshing my memory on SciPy's implementation: https://github.com/scipy/scipy/blob/3f11e32f3d07e6035b46f77eafec4359875a0416/scipy/fft/_helper.py#L6
and your suggestion

> (we can always add a note: "Due to differences in the underlying FFT libraries used by CuPy and SciPy, the length returned will not always be identical to scipy.fft.next_fast_len.")

brings up a question: with the recent backend support in `scipy.fft`, perhaps `scipy.fft.next_fast_len()` should also be registered with a backend? Ideally, we want this usage
```python
import scipy.fft as spfft
import cupyx.scipy.fft as cufft

spfft.set_global_backend(cufft)
n = spfft.next_fast_len(100)
```
to correctly return the number for cuFFT, not for pocketfft.

**Comment by leofang** [comment]
@asi1024 @grlee77 I don't think I fully understand this, but here you go:
```python 
cpdef size_t next_fast_len(size_t n):
    """2^a * 3^b * 5^c * 7^d for cuFFT."""

    if n <= 8:
        return n

    cdef size_t b = 1, c = 1, d = 1, best = 2 * n

    while d < best:
        c = d
        while c < best:
            b = c
            while b < n:
                b *= 2
            while True:
                if b < n:
                    b *= 3
                elif b > n:
                    if b < best:
                        best = b
                    if b & 1:
                        break
                    b >>= 1
                else:
                    return n
            c *= 5
        d *= 7
    return best
```

**Comment by grlee77** [comment]
> @asi1024 @grlee77 How about this? This is basically what SciPy does, tailored to primes (2, 3, 5, 7).

Yes, this looks correct. I empirically verified via comparison to `scipy.fft.next_fast_len` by checking that when there are factors of 11 the scipy result is smaller than this one, otherwise they are the same. details below (with `prime_factors` function taken from this [stackoverflow post](https://stackoverflow.com/questions/15347174/python-finding-prime-factors))

<details>

```Python
def prime_factors(n):
    i = 2
    factors = []
    while i * i <= n:
        if n % i:
            i += 1
        else:
            n //= i
            factors.append(i)
    if n > 1:
        factors.append(n)
    return factors


def next_fast_len(n):
    """2^a * 3^b * 5^c * 7^d for cuFFT."""
    if n <= 8:
        return n
    b = 1
    c = 1
    d = 1
    best = 2 * n
    while d < best:
        c = d
        while c < best:
            b = c
            while b < n:
                b *= 2
            while True:
                if b < n:
                    b *= 3
                elif b > n:
                    if b < best:
                        best = b
                    if b & 1:
                        break
                    b >>= 1
                else:
                    return n
            c *= 5
        d *= 7
    return best


for n in range(1024):
    n_fast_scipy = scipy.fft.next_fast_len(a)
    n_fast = next_fast_len(a)
    if 11 in prime_factors(n_fast_scipy):
        assert n_fast_scipy < n_fast
    else:
        assert n_fast_scipy == n_fast
```

</details>

**Comment by leofang** [comment]
Awesome! Thanks for confirming, @grlee77. 

@asi1024, would you like to include `next_fast_len()` in this PR, or should I send a new one? Note that it better be implemented in C or Cython, because it'd be way faster:
```python
In [3]: %timeit nfl_cython(100000000000)                                                                                                                             
489 ns ± 6.36 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: %timeit nfl_python(100000000000)                                                                                                                               
53.4 µs ± 787 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [6]: %timeit nfl_cython(1000000)                                                                                                                                    
176 ns ± 0.897 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [7]: %timeit nfl_python(1000000)                                                                                                                                    
16.2 µs ± 152 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

**Comment by asi1024** [comment]
PTAL.

**Comment by leofang** [comment]
@asi1024 Could you rerun your benchmark and see it's better? 😇

**Comment by asi1024** [comment]
Should we fix to 1. skip tests if `cupy.cuda.cufft.getVersion() == 10000` or 2. relax tolerance to pass with any version of cuFFT?

**Comment by leofang** [comment]
I am fine either way, was just wondering why we didn't see this issue in individual fft tests. 

@anaruse Do you happen to know if there's known precision issues for 1D FFT in CUDA 10.0?

**Comment by asi1024** [comment]
Updated performance numbers in the PR description!

**Comment by emcastillo** [comment]
I measured the dot convolution and the fft depending on the inputs a and b sizes. What I got is this

image

Y axis is a size, X axis is b size.
In green you can see the results when direct convolution is faster than fft.
This is measured in a V100 GPU so the threshold is likely to change depending on the GPU.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
@asi1024 @grlee77 I am wondering for `C2C`  if we could further speed up this by reusing a `Plan1d` for the three fft calls

**Comment by emcastillo** [comment]
Probably yes, most of the time is CPU bounded due to the plan creation. Can you do some test?

Thanks!

**Pull request merged**